### ISSUE TITLE
Style manipulation not robust, very sensitive to varied parameters

### ISSUE THREAD

**Issue by jiaoxijie** [LynnHo/AttGAN-Tensorflow#47]
How do you get a good balance among varied hyper-parameters like different loss weights, learning_rate when style manipulation is adopted? I found the training of the network very unstable.

I can get style manipulation results on bangs and eyeglasses, but the control is unstable and the sharpness of images are also affected. The control on eyeglasses is only on the shade and the model has no control on shape and size. 

Except for hyper-parameters, is there any other places where there can be problems like training settings? 

Besides, when implementing style manipulation, except for the loss of generated style controller, do you also use the original attribute loss?

Looking forward to your answer. Thank you!

**Comment by LynnHo** [comment]
@jiaoxijie I am sorry that I am too busy these days to answer your questions.

Your questions are not easy to answer because many situations cause your problems. First, what do you mean by "the control is unstable"? Second, as for the image quality, the training dynamic and the network architecture may affect it, if the training is stable, you can try some other architectures, e.g. CycleGAN generator. Third, I think you can try discreet variable to control eyeglasses, maybe it's more suitable to learn eyeglasses type.

"except for the loss of generated style controller, do you also use the original attribute loss?" Yes.

**Comment by XijieJiao** [comment]
Hi. Thank you so much for your reply. I have been struggling with it for weeks, so I understand that it is a complex problem.

'The control is unstable' means that multi-style exhibits only on some images. On others, the multi-styles are the same(or even the attribute control loses efficacy, but in your initial ATTGAN, the attribute control works well). Besides, the style manipulation on eyeglasses is only on the darkness of sunglasses. The Shape and the size of eyeglasses are not affected at all (I think the reason is the darkness of sunglasses is easier to learn than shape and size).

Second, I directly incorporate the CFGAN part into your ATTGAN which means that I added the extension of latent code from the existing labels (like CFGAN) and added the loss of the generated style controller.  I found that if not changing anything of your learning-rate, number of attributes used(13), and the weight of different losses, the style manipulation would not function at all. Right now, for the best result of mine, I increased the weight of style controller loss (but remove the original attribute loss as I stated below), increased the learning rate from 0.0002 to 0.0003, and even tried cyclical learning rate (because I found the model would be stuck in local optimum point which means that the details of images were recovered perfectly but the attributes were not changed or the styles were not extended at all. The losses also become stable too quickly). For the extended styles, I can only use four because the training becomes very hard when the number of styles increases to 6 or more. 
Did you also face that many problems when implementing your style manipulation part? Since it can work (though bad) on some images, I think the logic of my code should have no problems except for some other changes of ATTGAN needed when style manipulation is added.

Third. Well, that is different from CFGAN since CFGAN has only the loss of generated style controller and the original labels are only used for creating extended latent code. However, when I first added the style manipulation part, I did not remove your original attribute loss. When training, I found a great unbalance of the two attribute losses.  And in most cases, only the original attribute loss works (the loss of generated style controller declines and stabiles too quickly within one to two epochs).

Sorry for too many words, yet I have really done a lot of experiments and still cannot solve this problem.

I have one more question. I found most of the GAN papers do not even consider style manipulation. Is it itself a very tricky problem that it can work but cannot work perfectly as the many attribute editing tasks.
